### PR TITLE
EzDefi Extension

### PR DESCRIPTION
**Pull request by vuducchinh29**

### CODE REVIEW

File: src/utils/ezWallet.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
+import EventEmitter from 'eventemitter3';
```
Comment (Zergity) [inline src/utils/ezWallet.js]: Please don't copy the whole wallet here. import it from ezDeFi project

File: src/utils/markets.js [hunk 2]
```diff
@@ -50,18 +50,18 @@ export function useAllMarkets(customMarkets) {
             programId: marketInfo.programId,
           });
         } catch (e) {
-          notify({
-            message: 'Error loading all market',
-            description: e.message,
-            type: 'error',
-          });
+          // notify({
```
Comment (Zergity) [inline src/utils/markets.js]: please remove commented code

File: src/components/TopBar.js [hunk 3]
```diff
@@ -34,11 +34,18 @@ const LogoWrapper = styled.div`
 `;
 
 export default function TopBar() {
-  const { connected, wallet, providerUrl, setProvider } = useWallet();
+  const { connected, wallet, providerUrl, setProvider, cluster } = useWallet();
   const { endpoint, setEndpoint } = useConnectionConfig();
   const location = useLocation();
   const history = useHistory();
 
+  // Auto set endpoint when connected to ezDefi extension, available when testnet market is enabled
```
Comment (Zergity) [inline src/components/TopBar.js]: please remove commented code

File: src/utils/wallet.js [hunk 4]
```diff
@@ -42,6 +57,9 @@ export function WalletProvider({ children }) {
         message: 'Wallet update',
         description: 'Connected to wallet ' + keyToDisplay,
       });
+      wallet
```
Comment (Zergity) [inline src/utils/wallet.js:70]: I'm not sure if we need to handle cluster here, it seems that serum-dex only use the wallet adapter for message/tx signing, not sending. So it does not matter which cluster ezWallet is connecting to.

File: src/utils/wallet.js [hunk 5]
```diff
@@ -18,15 +20,28 @@ export function WalletProvider({ children }) {
     'https://www.sollet.io',
   );
 
-  const wallet = useMemo(() => new Wallet(providerUrl, endpoint), [
+  let wallet = useMemo(() => new Wallet(providerUrl, endpoint), [
     providerUrl,
     endpoint,
   ]);
+  const [ezWallet] = useState(new EZWallet());
```
Comment (Zergity) [inline src/utils/wallet.js]: put EZWallet initialization code in the const wallet = useMemo(...)

Comment (Zergity) [reply src/utils/wallet.js]: something like this
```
const wallet = useMemo(() => {
  if (new URL(providerUrl).hostname === 'ezdefi.com') {
    return new EzWallet();
  }
  return new Wallet(providerUrl, endpoint);
}, [
    providerUrl,
    endpoint,
  ]);
```

File: src/utils/wallet.js [hunk 6]
```diff
@@ -3,9 +3,11 @@ import Wallet from '@project-serum/sol-wallet-adapter';
 import { notify } from './notifications';
 import { useConnectionConfig } from './connection';
 import { useLocalStorageState } from './utils';
+import EZWallet from './ezWallet';
 
 export const WALLET_PROVIDERS = [
   { name: 'sollet.io', url: 'https://www.sollet.io' },
+  { name: 'ezDefi', url: 'https://ezdefi.com/wallet' },
```
Comment (Zergity) [inline src/utils/wallet.js]: please check brand styling, ezDefi or ezDeFi

File: src/utils/wallet.js [hunk 7]
```diff
@@ -3,9 +3,11 @@ import Wallet from '@project-serum/sol-wallet-adapter';
 import { notify } from './notifications';
 import { useConnectionConfig } from './connection';
 import { useLocalStorageState } from './utils';
+import EZWallet from './ezWallet';
```
Comment (Zergity) [inline src/utils/wallet.js]: import EzWallet from '@ezDeFi/ez-sol-wallet-adapter'

File: src/utils/markets.js [hunk 8]
```diff
@@ -50,18 +50,18 @@ export function useAllMarkets(customMarkets) {
...
+          //   type: 'error',
+          // });
         }
       }
       setMarkets(markets);
     };
 
     getAllMarkets();
-  }, [connection]);
+  }, [connection, customMarkets]);
```
Comment (Zergity) [inline src/utils/markets.js]: why there's `customMarkets` here?

Comment (vuducchinh29) [reply src/utils/markets.js]: I added it to remove the warning. Removed it.

File: src/utils/wallet.js (hunk 7 above)
Comment (vuducchinh29) [reply src/utils/wallet.js]: Resolved

File: src/utils/wallet.js (hunk 6 above)
Comment (vuducchinh29) [reply src/utils/wallet.js]: Resolved

File: src/utils/wallet.js (hunk 5 above)
Comment (vuducchinh29) [reply src/utils/wallet.js]: Tried it but it works incorrectly.
The default wallet was initialization is sollet.io. When switching the wallet, the wallet object is not re-initialization, an error: "wallet.on is not a function".

File: src/utils/wallet.js (hunk 4 above)
Comment (vuducchinh29) [reply src/utils/wallet.js:70]: Use cluster information to access explorer.solana.com in testnet/devnet/... to get account details.

File: src/components/TopBar.js (hunk 3 above)
Comment (vuducchinh29) [reply src/components/TopBar.js]: resolved

File: src/utils/markets.js (hunk 2 above)
Comment (vuducchinh29) [reply src/utils/markets.js]: resolved

File: src/utils/ezWallet.js (hunk 1 above)
Comment (vuducchinh29) [reply src/utils/ezWallet.js]: resolved

File: package.json [hunk 9]
```diff
@@ -8,6 +8,7 @@
     "@craco/craco": "^5.6.4",
     "@project-serum/serum": "0.13.1",
     "@project-serum/sol-wallet-adapter": "^0.1.1",
+    "@ezdefi/ez-sol-wallet-adapter": "git://github.com/ezdefi/ez-sol-wallet-adapter.git#0.1.1",
```
Comment (Zergity) [inline package.json]: choose something flexible for the tag name so we can update it later: `sereum` is a good choice, later we can keep the tag along with the latest branch

File: src/utils/ezWallet.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+import { PublicKey } from '@solana/web3.js';
+import bs58 from 'bs58';
+const ethereum = window.ethereum;
```
Comment (Zergity) [inline src/utils/ezWallet.js]: use `solana = window.solana` instead

File: src/utils/ezWallet.js (hunk 1 above)
Comment (Zergity) [inline src/utils/ezWallet.js]: should this file be removed?

**Review by Zergity — CHANGES_REQUESTED** [review]

File: src/utils/wallet.js [hunk 11]
```diff
@@ -18,16 +20,37 @@ export function WalletProvider({ children }) {
     'https://www.sollet.io',
   );
 
-  const wallet = useMemo(() => new Wallet(providerUrl, endpoint), [
+  let wallet = useMemo(() => new Wallet(providerUrl, endpoint), [
```
Comment (Zergity) [inline src/utils/wallet.js:25]: This part still has a problem of creating 2 wallet instances, and only 1 is used. Please check the condition, and create only Wallet or EzWallet.

File: src/utils/wallet.js [hunk 12]
```diff
@@ -18,16 +20,37 @@ export function WalletProvider({ children }) {
...
+  if (providerUrl === 'https://ezdefi.com/wallet') {
+    wallet = ezWallet;
+  }
 
   const [connected, setConnected] = useState(false);
+  const [cluster, setCluster] = useState(false);
 
   useEffect(() => {
     console.log('trying to connect');
+    if (wallet === ezWallet && !window.solana) {
```
Comment (Zergity) [inline src/utils/wallet.js]: only 1 minor change is recommended: move the `window.solana` check into ezWallet adapter function:
```
if (wallet.isInjected && !wallet.isInjected())
```

**Review by Zergity — CHANGES_REQUESTED** [review]

**Review by Zergity — APPROVED** [review]
LGTM

**Pull request merged**